pkg/rbd/nodeserver.go

```diff
@@ -103,16 +97,29 @@ func (ns *NodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
 	}
 
 	if !isNotMnt {
-		klog.Infof(util.Log(ctx, "rbd: volume %s is already mounted to %s, skipping"), req.GetVolumeId(), stagingTargetPath)
+		klog.Infof(util.Log(ctx, "rbd: volume %s is already mounted to %s, skipping"), volID, stagingTargetPath)
 		return &csi.NodeStageVolumeResponse{}, nil
 	}
 
+	isLegacyVolume := isLegacyVolumeID(volID)
 	volOptions, err := genVolFromVolumeOptions(ctx, req.GetVolumeContext(), req.GetSecrets(), disableInUseChecks, isLegacyVolume)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
-	volOptions.RbdImageName = volName
-	volOptions.VolID = req.GetVolumeId()
+
+	var vi util.CSIIdentifier
+	err = vi.DecomposeCSIID(volID)
+	if err != nil {
+		err = fmt.Errorf("error decoding volume ID (%s) (%s)", err, volID)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	_, volOptions.RbdImageName, _, err = volJournal.GetObjectUUIDData(ctx, volOptions.Monitors, cr, volOptions.Pool, vi.ObjectUUID, false)
```

Comment:
For legacy volumes the image name was the PV name [for v1.0 version](https://github.com/ceph/ceph-csi/blob/2623bce76156f6de6cebb06b57ddfcd6b4c4e807/pkg/rbd/controllerserver.go#L119-L124).

The code here would fail to return the right legacy image name, as the volume name would be returned as "csi-vol-<UUID>", rather than the PV name that was passed initially.

As we do not store the PV name the function `getLegacyVolumeName` was the workaround to determine the image name from the mount path, as this carries the PV name in the path.

The code above needs to perform the fetch of image name based on the OMap if this is not a legacy volume, else fall back to the said function to get the same.

Reply:
ack!